### PR TITLE
StevensonSpace Jukebox

### PR DESCRIPTION
**Pull request by aw-0**
Listen to unique, original music by Stevenson students.

- [x] Playing different songs
- [x] Skipping songs
- [x] Song submission form
- [x] Finish previous song behavior
- [x] Build Jukebox backend on CF Workers to ratelimit R2 access
- [x] Send Google Form out to music dept.
- [x] Add a card advertising that the feature is "Coming Soon", with GForm link
- [x] Once we get a few songs in the catalog, we go live!

Still WIP, but nearly there -- wanted to make a draft so I can get a preview URL to eventually share w/ the teacher who suggested this.

### CODE REVIEW

**Comment by aw-0** [comment]
Jukebox is almost ready to go live! Here's the roadmap:

- Finish song scrubbing (feel like this is a needed feature)
- Send Google Form out to music dept.
- Potentially add a card advertising that the feature is "Coming Soon", with GForm link? cc @JosephShepin 
- Once we get a few songs in the catalog, we go live!

**Comment by JosephShepin** [comment]
Ok nice. Where are we gonna store the songs?

**Comment by aw-0** [comment]
Songs are stored in /public/music/audio - same with album art @ /public/music/art

Then the browser just loads both when a song is played based on a .json file that has the song info on it

**Comment by JosephShepin** [comment]
This is a good question. I think it might be difficult to find/retain someone to do that especially since there's semester gaps between batches of submissions. I don't think we should have to be responsible for checking the music and providing feedback - we can listen through it to make sure but it should be expected of the teacher and students that the music is clean before it's handed to us. Our job is the software which you've already put lots of time and effort into and their job at minimum is providing us with unloadable music.

**Comment by aw-0** [comment]
I totally agree. I think that if the music dept. themselves want to appoint the student and manage all the associated work that comes with that, I'd be more than happy with just adding someone as a middle-man on the Google Form. But I don't really want to be responsible for finding someone, figuring out if they're credible, etc.

**Comment by Chaldron** [comment]
> Songs are stored in /public/music/audio - same with album art @ /public/music/art
> 
> Then the browser just loads both when a song is played based on a .json file that has the song info on it

Is there another place we could store these at perhaps some (low) cost? Storing binary files in git can blow up the `.git` folder since they will remain there in the history forever even if they're deleted in a commit. It can make checking out the repository a lot slower.

One solution to this is [Git LFS](https://git-lfs.com) and it seems like GitHub gives us 1 GB free storage and 1 GB free bandwidth per month. Since I'm guessing the media will be rotated in out and this might work although 1 GB might be a little low (e.g. a 3 MB file could be streamed/downloaded only ~300 times). Netlify also seems to have their own implementation of LFS but they don't list pricing clearly anywhere unless I'm missing it.

**Comment by JosephShepin** [comment]
@Chaldron can you bill the school for this? Also the domain.

**Comment by JosephShepin** [comment]
Git LFS seems like a good idea. If we go over we can explore free tiers for storage and cdn services.

**Comment by yvvijay121** [comment]
OK, this might be an idea that may have already been considered, but what if we used [SoundCloud's streaming API](https://developers.soundcloud.com/docs/api/guide#playing)? This would let people provide the songs easily through a SoundCloud link, and there would be no need to store the audio files. This is a completely naive suggestion, as I haven't looked into any of the cons of this solution.

(also, I'm in advanced music production right now, so I could help a little bit with this)

**Comment by aw-0** [comment]
That's definitely a good point I've never thought about. Git LFS *could* work, but GitHub's pricing is kinda high for it, and tbf Git really wasn't made to store this type of data.

SoundCloud could work too, my only main gripe with it is that it requires you to make an account with in order to use, which could add an extra step of friction to the form.

I might've found a solution. Cloudflare has a S3 cdn competitor called R2, and it's actually pretty cheap. 10 gb/month of free storage, 10 million requests/month of reap op's. Think instead of just storing the audio in the Git repo/website itself, we should store it in R2. There's also the additional benefit of having the cf network cache all of the songs and make it faster for all users.

**Comment by aw-0** [comment]
Just pushed a fix that moves all the audio from local to R2 - it actually works pretty great. slight delay in loading the songs the first time but then they're cached so pretty much instant loading times after that

**Comment by aw-0** [comment]
In order to prevent DDOS attacks, I've added a simple backend hosted on Cloudflare Workers that runs off of Upstash Serverless Redis to add rate limiting—additionally, I fixed a few minor issues as well

Really only thing that needs to get fixed is the mobile support, but after that, we should be ready to ship!

**Comment by aw-0** [comment]
And... tested this today - the rate limiter does not work on the school network as all devices share the same external IP. Going to spend the next little while thinking of the best way to fix this. I've came up with 2 options:

- Require a Google Account login, and use the uid from that as the rate limiter uid
- Stick an invisible captcha behind every request

Not entirely sure what the best solution to this will be.

**Comment by JosephShepin** [comment]
Screenshot 2023-01-24 at 12 24 07 AM
is there a way we ca use cloudflare's ddos protection?

**Comment by aw-0** [comment]
Spent the last little while thinking about how to fix this issue, and I'm not sure if there's a great solution either way. Here's what I propose:

We release Jukebox with no rate limiting, besides Cloudflare native DDOS protection. Hopefully, that will block a majority of malicious requests, but if there's an attack bigger than that, then we can ask CF for a reimbursement for the attack (have had to do that with GCP, AWS, and more and they're usually fine with it) and we lock Jukebox behind school GSuite accounts. That way we truly have a unique ID in the rate limiter's database, so it'll work fine.

Basically, I say we give the benefit of the doubt that no one tries to attack, and if they do, then we go ahead and lock everything down. Seems like the best way to do this 🤷 - lmk if you have any other better ideas @JosephShepin

**Comment by aw-0** [comment]
Merging Wednesday Night 🎉

**Comment by aw-0** [comment]
:tada:
